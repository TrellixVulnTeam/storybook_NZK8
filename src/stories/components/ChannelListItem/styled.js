// @flow

import styled from 'styled-components';
import { colors } from 'utils/StyleGuide';

export const Wrapper = styled.li`
  &:hover,
  &:hover div,
  &:hover label {
    cursor: ${p => (p.disablePointer ? 'auto' : 'pointer')};
  }

  &:hover {
    background-color: ${colors.lightGrey};
  }

  margin: 0;
  padding: 18px 20px 18px 18px;
  position: relative;
  display: flex;
  flex-flow: row nowrap;
  justify-content: flex-start;
  align-items: center;
  background-image: ${p => (p.active ? 'linear-gradient(253deg, #4da2ff, #0076ff)' : 'none')};
`;

export const Avatar = styled.img`
  width: 50px;
  height: 50px;
  border-radius: 100%;
`;

export const Participants = styled.div`
  &:hover {
  }

  width: 50px;
  height: 50px;
  border-radius: 100%;
  background-color: ${p => (p.active ? colors.white : colors.babyBlue)};
  display: flex;
  justify-content: center;
  align-items: center;
  font-size: 18px;
  font-weight: bold;
  color: ${colors.brightBlue};
`;

export const Details = styled.div`
  flex: 1;
  padding-left: 12px;
  width: 100%;
  overflow: hidden;
`;

export const Title = styled.div`
  display: flex;
  flex-flow: row nowrap;
  justify-content: space-between;
  align-items: flex-start;

  label:first-child {
    display: block;
    font-size: 16px;
    color: ${p => (p.active ? colors.white : colors.slateGrey)};
    font-weight: 600;
    text-overflow: ellipsis;
    overflow: hidden;
  }

  label:last-child {
    font-size: 10px;
    color: ${p => (p.active ? colors.white : p.slateGrey)};
  }
`;

export const Body = styled.div`
  display: block;
  font-size: 12px;
  color: ${p => (p.active ? colors.white : colors.steel)};
  font-weight: ${p => (p.unread ? 'bold' : 'normal')};
`;

export const Date = styled.div`
  align-self: flex-start;
  font-size: 10px;
  color: ${colors.slateGrey};
`;

export const Star = styled.span`
  position: absolute;
  top: 0;
  right: 2px;
  color: ${colors.marigold};
`;

export const Unread = styled.span`
  position: absolute;
  top: 0;
  right: 2px;
  color: ${colors.strawberry};
  font-size: 45px;
  line-height: 45px;
`;
