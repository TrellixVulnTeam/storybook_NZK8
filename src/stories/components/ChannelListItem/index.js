// @flow

import React from 'react';
import get from 'lodash.get';
import moment from 'moment';
import { ChatMarkdown } from 'utils';
import { Wrapper, Avatar, Details, Title, Body, Participants, Star, Unread } from './styled';

type Props = {
  channel: {
    id: number | string,
    type: 'DIRECT_MESSAGE' | 'GROUP_MESSAGE' | 'WORK_ORDER',
    users: Array<Object>,
    pinned: boolean,
    name: string,
    avatar?: ?string,
    lastChannelMessage: Object,
  },
  active: boolean,
  unread: boolean,
  session: Object,
  onClick: Object => void,
};

export default (props: Props) => {
  const messageBody = get(props, 'channel.lastChannelMessage.body', '');
  const messagePerson = get(props, 'channel.lastChannelMessage.user.firstName', '');
  const date = get(props, 'channel.lastChannelMessage.createdAt');

  let message = ChatMarkdown.stripMarkdown(
    (messagePerson ? `${messagePerson}: ` : '') + messageBody,
  );
  if (message.length > 155) {
    message = `${message.substr(0, 40)}...`;
  }

  const otherUser = props.channel.users.find(a => a.id !== props.session.id);
  const otherUserAvatar =
    props.channel.type === 'DIRECT_MESSAGE' && otherUser
      ? otherUser.avatar
      : '/static/web/assets/defaultAvatar.png';

  let channelName = props.channel.name;
  if (props.channel.type === 'DIRECT_MESSAGE' && otherUser) {
    if (otherUser.firstName && otherUser.lastName) {
      channelName = `${otherUser.firstName} ${otherUser.lastName}`;
    } else {
      channelName = otherUser.email;
    }
  }

  return (
    <Wrapper
      pinned={props.channel.pinned}
      onClick={() => props.onClick(props.channel)}
      active={props.active}
    >
      {props.channel.type === 'DIRECT_MESSAGE' ? (
        <Avatar src={otherUserAvatar || '/static/web/assets/defaultAvatar.png'} />
      ) : (
        <Participants active={props.active}>{props.channel.users.length}</Participants>
      )}
      <Details>
        <Title active={props.active}>
          {/* eslint-disable-next-line jsx-a11y/label-has-associated-control */}
          <label>{channelName}</label>
          {/* eslint-disable-next-line jsx-a11y/label-has-associated-control */}
          <label>{moment(date).fromNow()}</label>
        </Title>
        <Body active={props.active} unread={props.unread}>
          {message}
        </Body>
      </Details>
      {props.channel.pinned ? <Star>⭑</Star> : null}
      {props.unread ? <Unread>·</Unread> : null}
    </Wrapper>
  );
};
