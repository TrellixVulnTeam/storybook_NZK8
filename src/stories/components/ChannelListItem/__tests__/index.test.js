import React from 'react';
import { mount } from 'enzyme';
import ChannelListItem from 'components/ChannelListItem/index';

describe('<ChannelListItem />', () => {
  const onClickMock = jest.fn();
  const mandatoryProps = {
    channel: {
      id: 0,
      type: 'DIRECT_MESSAGE',
      users: [
        {
          id: 1,
        },
      ],
      pinned: true,
      avatar: 'Avatar',
      lastChannelMessage: {
        body: '',
        user: {
          firstname: 'Name',
        },
        createdAt: '',
      },
    },
    active: true,
    unread: true,
    session: {
      id: 2,
    },
    onClick: onClickMock,
  };

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('should render correctly with mandatory props', () => {
    const sut = mount(<ChannelListItem {...mandatoryProps} />);

    expect(sut).toMatchSnapshot();
  });

  it('should render correctly for not direct messages', () => {
    const props = {
      ...mandatoryProps,
      channel: {
        ...mandatoryProps.channel,
        type: 'GROUP_MESSAGE',
      },
    };
    const sut = mount(<ChannelListItem {...props} />);

    expect(sut).toMatchSnapshot();
  });

  it('should render correctly for inactive not direct messages', () => {
    const props = {
      ...mandatoryProps,
      channel: {
        ...mandatoryProps.channel,
        type: 'GROUP_MESSAGE',
      },
      active: false,
    };
    const sut = mount(<ChannelListItem {...props} />);

    expect(sut).toMatchSnapshot();
  });

  it(`should render correctly for inactive, read, body message more than 155 chars
    and user with firstname and lastname`, () => {
    const props = {
      ...mandatoryProps,
      channel: {
        id: 0,
        type: 'DIRECT_MESSAGE',
        users: [
          {
            id: 1,
            firstName: 'First',
            lastName: 'Last',
          },
          {
            id: 2,
            firstName: 'First',
            lastName: 'Last',
          },
          {
            id: 3,
            firstName: 'First',
            lastName: 'Last',
          },
        ],
        pinned: false,
        name: 'Name',
        avatar: 'Avatar',
        lastChannelMessage: {
          body: `Should be more than 155 chars to test a specific case
            Should be more than 155 chars to test a specific case
            Should be more than 155 chars to test a specific case.`,
          user: {
            firstName: 'Name',
            lastName: 'Lastname',
          },
          createdAt: '',
        },
      },
      active: false,
      unread: false,
      session: {
        id: 2,
      },
    };
    const sut = mount(<ChannelListItem {...props} />);

    expect(sut).toMatchSnapshot();
  });

  it('should call custom click handler', () => {
    const sut = mount(<ChannelListItem {...mandatoryProps} />);

    expect(onClickMock).toHaveBeenCalledTimes(0);

    sut.simulate('click');

    expect(onClickMock).toHaveBeenCalledTimes(1);
  });
});
