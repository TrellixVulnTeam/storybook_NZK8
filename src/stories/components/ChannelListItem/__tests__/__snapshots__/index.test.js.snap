// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ChannelListItem /> should render correctly for inactive not direct messages 1`] = `
<_default
  active={false}
  channel={
    Object {
      "avatar": "Avatar",
      "id": 0,
      "lastChannelMessage": Object {
        "body": "",
        "createdAt": "",
        "user": Object {
          "firstname": "Name",
        },
      },
      "pinned": true,
      "type": "GROUP_MESSAGE",
      "users": Array [
        Object {
          "id": 1,
        },
      ],
    }
  }
  onClick={[MockFunction]}
  session={
    Object {
      "id": 2,
    }
  }
  unread={true}
>
  <styled__Wrapper
    active={false}
    onClick={[Function]}
    pinned={true}
  >
    <li
      className="styled__Wrapper-sc-1wjrk8z-0 ljWMgc"
      onClick={[Function]}
    >
      <styled__Participants
        active={false}
      >
        <div
          className="styled__Participants-sc-1wjrk8z-2 koGZMK"
        >
          1
        </div>
      </styled__Participants>
      <styled__Details>
        <div
          className="styled__Details-sc-1wjrk8z-3 ZkCsN"
        >
          <styled__Title
            active={false}
          >
            <div
              className="styled__Title-sc-1wjrk8z-4 bjYnLH"
            >
              <label />
              <label>
                Invalid date
              </label>
            </div>
          </styled__Title>
          <styled__Body
            active={false}
            unread={true}
          >
            <div
              className="styled__Body-sc-1wjrk8z-5 jxxAhQ"
            />
          </styled__Body>
        </div>
      </styled__Details>
      <styled__Star>
        <span
          className="styled__Star-sc-1wjrk8z-7 gsAcqq"
        >
          ⭑
        </span>
      </styled__Star>
      <styled__Unread>
        <span
          className="styled__Unread-sc-1wjrk8z-8 eTlaFo"
        >
          ·
        </span>
      </styled__Unread>
    </li>
  </styled__Wrapper>
</_default>
`;

exports[`<ChannelListItem /> should render correctly for inactive, read, body message more than 155 chars
    and user with firstname and lastname 1`] = `
<_default
  active={false}
  channel={
    Object {
      "avatar": "Avatar",
      "id": 0,
      "lastChannelMessage": Object {
        "body": "Should be more than 155 chars to test a specific case
            Should be more than 155 chars to test a specific case
            Should be more than 155 chars to test a specific case.",
        "createdAt": "",
        "user": Object {
          "firstName": "Name",
          "lastName": "Lastname",
        },
      },
      "name": "Name",
      "pinned": false,
      "type": "DIRECT_MESSAGE",
      "users": Array [
        Object {
          "firstName": "First",
          "id": 1,
          "lastName": "Last",
        },
        Object {
          "firstName": "First",
          "id": 2,
          "lastName": "Last",
        },
        Object {
          "firstName": "First",
          "id": 3,
          "lastName": "Last",
        },
      ],
    }
  }
  onClick={[MockFunction]}
  session={
    Object {
      "id": 2,
    }
  }
  unread={false}
>
  <styled__Wrapper
    active={false}
    onClick={[Function]}
    pinned={false}
  >
    <li
      className="styled__Wrapper-sc-1wjrk8z-0 ljWMgc"
      onClick={[Function]}
    >
      <styled__Avatar
        src="/static/web/assets/defaultAvatar.png"
      >
        <img
          className="styled__Avatar-sc-1wjrk8z-1 bvoqYD"
          src="/static/web/assets/defaultAvatar.png"
        />
      </styled__Avatar>
      <styled__Details>
        <div
          className="styled__Details-sc-1wjrk8z-3 ZkCsN"
        >
          <styled__Title
            active={false}
          >
            <div
              className="styled__Title-sc-1wjrk8z-4 bjYnLH"
            >
              <label>
                First Last
              </label>
              <label>
                Invalid date
              </label>
            </div>
          </styled__Title>
          <styled__Body
            active={false}
            unread={false}
          >
            <div
              className="styled__Body-sc-1wjrk8z-5 dgzbXo"
            >
              Name Should be more than 155 chars to te...
            </div>
          </styled__Body>
        </div>
      </styled__Details>
    </li>
  </styled__Wrapper>
</_default>
`;

exports[`<ChannelListItem /> should render correctly for not direct messages 1`] = `
<_default
  active={true}
  channel={
    Object {
      "avatar": "Avatar",
      "id": 0,
      "lastChannelMessage": Object {
        "body": "",
        "createdAt": "",
        "user": Object {
          "firstname": "Name",
        },
      },
      "pinned": true,
      "type": "GROUP_MESSAGE",
      "users": Array [
        Object {
          "id": 1,
        },
      ],
    }
  }
  onClick={[MockFunction]}
  session={
    Object {
      "id": 2,
    }
  }
  unread={true}
>
  <styled__Wrapper
    active={true}
    onClick={[Function]}
    pinned={true}
  >
    <li
      className="styled__Wrapper-sc-1wjrk8z-0 jIiWfe"
      onClick={[Function]}
    >
      <styled__Participants
        active={true}
      >
        <div
          className="styled__Participants-sc-1wjrk8z-2 fkmtTt"
        >
          1
        </div>
      </styled__Participants>
      <styled__Details>
        <div
          className="styled__Details-sc-1wjrk8z-3 ZkCsN"
        >
          <styled__Title
            active={true}
          >
            <div
              className="styled__Title-sc-1wjrk8z-4 kjaPRJ"
            >
              <label />
              <label>
                Invalid date
              </label>
            </div>
          </styled__Title>
          <styled__Body
            active={true}
            unread={true}
          >
            <div
              className="styled__Body-sc-1wjrk8z-5 kRivHZ"
            />
          </styled__Body>
        </div>
      </styled__Details>
      <styled__Star>
        <span
          className="styled__Star-sc-1wjrk8z-7 gsAcqq"
        >
          ⭑
        </span>
      </styled__Star>
      <styled__Unread>
        <span
          className="styled__Unread-sc-1wjrk8z-8 eTlaFo"
        >
          ·
        </span>
      </styled__Unread>
    </li>
  </styled__Wrapper>
</_default>
`;

exports[`<ChannelListItem /> should render correctly with mandatory props 1`] = `
<_default
  active={true}
  channel={
    Object {
      "avatar": "Avatar",
      "id": 0,
      "lastChannelMessage": Object {
        "body": "",
        "createdAt": "",
        "user": Object {
          "firstname": "Name",
        },
      },
      "pinned": true,
      "type": "DIRECT_MESSAGE",
      "users": Array [
        Object {
          "id": 1,
        },
      ],
    }
  }
  onClick={[MockFunction]}
  session={
    Object {
      "id": 2,
    }
  }
  unread={true}
>
  <styled__Wrapper
    active={true}
    onClick={[Function]}
    pinned={true}
  >
    <li
      className="styled__Wrapper-sc-1wjrk8z-0 jIiWfe"
      onClick={[Function]}
    >
      <styled__Avatar
        src="/static/web/assets/defaultAvatar.png"
      >
        <img
          className="styled__Avatar-sc-1wjrk8z-1 bvoqYD"
          src="/static/web/assets/defaultAvatar.png"
        />
      </styled__Avatar>
      <styled__Details>
        <div
          className="styled__Details-sc-1wjrk8z-3 ZkCsN"
        >
          <styled__Title
            active={true}
          >
            <div
              className="styled__Title-sc-1wjrk8z-4 kjaPRJ"
            >
              <label />
              <label>
                Invalid date
              </label>
            </div>
          </styled__Title>
          <styled__Body
            active={true}
            unread={true}
          >
            <div
              className="styled__Body-sc-1wjrk8z-5 kRivHZ"
            />
          </styled__Body>
        </div>
      </styled__Details>
      <styled__Star>
        <span
          className="styled__Star-sc-1wjrk8z-7 gsAcqq"
        >
          ⭑
        </span>
      </styled__Star>
      <styled__Unread>
        <span
          className="styled__Unread-sc-1wjrk8z-8 eTlaFo"
        >
          ·
        </span>
      </styled__Unread>
    </li>
  </styled__Wrapper>
</_default>
`;
